### ISSUE TITLE
Add tests in sample Android app for most of Robolectric features using RoboSpock

### ISSUE THREAD

**Issue by pjakubczyk** [robospock/RoboSpock#3]
Copied from:
https://github.com/Polidea/RoboSpock/issues/5

**Comment by wojtekerbetowski** [comment]
You are right. That's gonna be priority no 1 now

**Comment by futtetennista** [comment]
Have you guys any idea of how much work it'll be? I'd be glad to help.

**Comment by wojtekerbetowski** [comment]
I think within next two weeks. Soon I'll push a branch for that. Feel free to ask and contribute.
This is mostly a rewrite, as we're using Robolectric's internals and those changed deeply

**Comment by futtetennista** [comment]
Yeah I quickly tried to use Robolectric 2.1 and saw a bunch of compilation errors coming up. Let me know when that branch is pushed.

**Comment by wojtekerbetowski** [comment]
Unfortunately the abstraction of Spock is in conflict with Robolectric classloader changing per test.

After 2 days of investigation I may say, that we simply won't be able to update to Robolectric 2.1
without compromising some of the Robolectric features :-(

By now it will look like following:

We still need to override the spock.lang.Specification to switch the classloader in the Sputnik test runner (just like before). By that time we need to capture the Robolectric configuration from file and/or class @Config annotation (new Robolectric feature).

Next, we need to create the classloader (some changes in Robolectric builder are required as it's being built per method).

Additionally we should add a check for using the @Config annotation in methods and raise a proper exception, so that wouldn't be confusing when it not works.

**Comment by wojtekerbetowski** [comment]
Check out branch https://github.com/Polidea/RoboSpock/tree/robolectric-2.1

This is going to be harder than I expected. At the moment in branch the Sputnik Test Runner is being created, but faces asm exception while constructing ClassLoader (probably conflicts with Groovy using ASM as well).

```
java.lang.IncompatibleClassChangeError: class org.objectweb.asm.tree.ClassNode has interface org.objectweb.asm.ClassVisitor as super class 
```

@pjakubczyk @stefanodacchille feel free to take over. Any ideas are welcome!

**Comment by futtetennista** [comment]
I'll have a look at it asap, I have to understand how spock works internally - just discovered it a few day ago. I run the groovy tests and I get a different error: `
org.spockframework.runtime.InvalidSpecException: Class 'spock.lang.Specification' is not a Spock specification (does not extend spock.lang.Specification or a subclass thereof)`

Update: I see the exception you mentioned if I run the tests from the command line. But in IntelliJ it the exception is different.

**Comment by wojtekerbetowski** [comment]
Sneak peek - code base is updated to Robolectric 2.1 (working PoC). Next I need to update the test project and documentation. After that I will release the new version of RoboSpock.

**Comment by futtetennista** [comment]
This sounds wicked! Could you push the latest changes in the robolectric-2.1 branch so that I can give it a try?

**Comment by futtetennista** [comment]
Just noticed you pushed it already to master. Will give it a try.

Could you maybe shorty explain what was the problem?

**Comment by wojtekerbetowski** [comment]
The main problem is that robolectric creates (+- caching) classloader per method, as each method might have a different Config annotation. This is not only the problem in migrating to Spock, but also doesn't (and won't) work with JUnit (static shared fields, `@BeforeClass` annotations).

Therefore I dropped the support, which made it possible to migrate the code in `RobolectricTestRunner` more easily. In upcoming days I should push more tests and examples and update the docs.

**Comment by wojtekerbetowski** [comment]
Some last fixes for Robolectric itself and the migration is mostly done. I need to update the docs and tests as well. Also providing some test apps would help. And last - revive RoboGuice support.
But I'm creating new issues for that.
